### PR TITLE
Add network policy for RKE2 Ingress Nginx

### PR DESCRIPTION
**Pull request by dharmit**
## Issue: 

Addresses issue raised in https://github.com/rancher/rancher/issues/45603.

## Problem

Prometheus target failing to come up

## Solution

Using the modified chart in this PR enables the target to come up

## Testing

- Install rancher-monitoring-crd through this PR
- Install rancher-monitoring using following Helm command:
  ```sh
  helm install rancher-monitoring charts/rancher-monitoring/105.0.0-rc1+up57.0.3/ \
  --namespace=cattle-monitoring-system \
  --set alertmanager.alertmanagerSpec.configSecret=alertmanager-rancher-monitoring-alertmanager \
  --set alertmanager.alertmanagerSpec.useExistingSecret=true \
  --set global.cattle.clusterId=local \
  --set global.cattle.clusterName=local \
  --set global.cattle.rkePathPrefix="" \
  --set global.cattle.rkeWindowsPathPrefix="" \
  --set global.cattle.systemDefaultRegistry="" \
  --set global.cattle.systemProjectId=p-cwh9d \
  --set global.cattle.url=https://13.202.137.42.sslip.io \
  --set global.systemDefaultRegistry="" \
  --set prometheus.prometheusSpec.evaluationInterval=1m \
  --set prometheus.prometheusSpec.retentionSize=50GiB \
  --set prometheus.prometheusSpec.scrapeInterval=1m \
  --set rke2ControllerManager.enabled=true \
  --set rke2Etcd.enabled=true \
  --set rke2IngressNginx.enabled=true \
  --set rke2IngressNginx.networkPolicy.enabled=true \
  --set rke2Proxy.enabled=true \
  --set rke2Scheduler.enabled=true
  ````
- Go to Rancher UI; Monitoring; Prometheus Targets
- In the Prometheus UI search for the ServiceMonitor `serviceMonitor/kube-system/rancher-monitoring-ingress-nginx/0`. It should be `UP`.

### CODE REVIEW

**Comment by dharmit** [comment]
I have honestly got no clue about why there are so many lines modified in this PR. Steps I followed are below:
```sh
$ export PACKAGE=rancher-monitoring/rancher-monitoring
$ make prepare
// made changes
$ make patch
$ make charts
```
Did I do something wrong here? 🤔

**Review by alexandreLamarre — DISMISSED** [review]
Rancher monitoring version needs to be bumped to `104.1.1-rc1`. Recommend separating `make patch` into a separate commit from `make charts` for exactly this scenario, where you would to do something like rebase with drop.

Need to remove `packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/crds` as a bug in charts-build-scripts includes these by accident.

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/crds/crds/crd-alertmanagerconfigs.yaml
Comment (joshmeranda) [inline packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/crds/crds/crd-alertmanagerconfigs.yaml:1]: Deleting this `packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/crds` directory should clean up those excess lines. This is a known issue caused by how we need to pull in our crds. There is a fix pending: https://github.com/rancher/charts-build-scripts/pull/143

**Review by joshmeranda — CHANGES_REQUESTED** [review]

**Comment by dharmit** [comment]
@joshmeranda @alexandreLamarre thanks for your feedback, guys! I have modified as per your recommendations. Can you PTAL again? Thanks.

**Comment by joshmeranda** [comment]
Looking better, but git histroy could use some housekeeping. Can you update the commits to look closer to the steps outline [here](https://github.com/rancher/charts/blob/dev-v2.9/docs/developing.md#making-changes-to-packages)? Generally this will look like this:

**Review by joshmeranda — CHANGES_REQUESTED** [review]
Looking better, but git history could use some housekeeping. Can you update the commits to look closer to the steps outline [here](https://github.com/rancher/charts/blob/dev-v2.9/docs/developing.md#making-changes-to-packages)? Generally this will look like this:

1. `export PACKAGE=rancher-monitoring/rancher-monitoring`
2. `make prepare`
3. Make the changes you need to the prepare chart, in this case adding the network policy
4. `make patch`
5. Commit the changes to `packages/rancher-monitoring/rancher-monitoring/generated-changes` (don't forget to remove `generated-changes/overlay/crds` introduced by the bug)
6. Update the version in `package.yaml`
7. Commit changes to `package.yaml`
8. `make charts`
9. Commit changes made by `make charts`
10. Update and commit `release.yaml` with new version of chart

Following these steps will make it a good bit easier for the reviewer to see what was changed and why. Also make each commit smaller and prevents new chart versions from stalling our browsers when trying to load 100+ new files. Thanks!

**Comment by dharmit** [comment]
@joshmeranda I referred to the [Making Changes To Packages](https://github.com/rancher/charts/blob/dev-v2.9/docs/developing.md#making-changes-to-packages) section which mentions making version related changes first. In context of your comment, I followed steps 6-9 first and 1-5 after that.

If I follow the steps you suggested, it'd make changes to 104.1.0. And, AFAIK, 104.1.0 has already released, right? Correct me if I'm wrong because there's a good chance I am. :)

**Comment by joshmeranda** [comment]
@dharmit You're right that we don't want to make changes to `104.1.0` but since you won't run `make charts` until after you bump the the version in step 6/7 you shouldn't be updating any existing charts.

I keep forgetting that the docs recommend you update the version first. That's fine as long as its a separate commit. Generally I'll update `package.yaml` in its own commit since its easier to see the new version in a commit with 1 changed file rather than 100+ changes files, then run `make charts` at the end. This will still create the new charts at versions `104.1.1` since `package.yaml` will have the bumped version.

File: index.yaml
Comment (ericpromislow) [inline index.yaml]: There's an extra space before `url:`. Strictly speaking it's not a yaml error, as this block is a string, but it looks like it's intended to be input for some other yaml process, and therefore should be conforming yaml code

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Something in this process is generating broken yaml.

**Comment by dharmit** [comment]
> I keep forgetting that the docs recommend you update the version first. That's fine as long as its a separate commit. Generally I'll update `package.yaml` in its own commit since its easier to see the new version in a commit with 1 changed file rather than 100+ changes files, then run `make charts` at the end. This will still create the new charts at versions `104.1.1` since `package.yaml` will have the bumped version.

I see your point. However, what I have done is again just as per the docs recommendation. It recommends updating the version in `package.yaml`, then doing `make charts`, and committing the changes.

I can totally do what you're recommending if that's the general practice followed by O&B team. Let me know.

**Comment by joshmeranda** [comment]
@dharmit Yeah that's totally fine to do it that way, its just less familiar. It makes it easier to ensure that the changes you made were brought into the generated chart. Feel free to continue!

But you'll still want to have separate commits for updating `package.yaml`, building the new charts at that version, making the necessary changes, and rebuilding the charts. The commit history for something like this pr should probably look like this:

1. Commit 1: `bump rancher-monitoring/rancher-monitoring version to 104.1.1+up57.0.3`
2. Commit 2: `make charts`
3. Commit 3: `update release.yaml` (can go anywhere as long as its updated`
4. Commit 4: `Add network policy for RKE2 Nginx Ingress`
5. Commit 5: 'make charts'

I get that this feels somewhat nitpicky, but because of the sheer amount of files that are updated in charts PRs it makes it much easier for reviewers to give a review.

**Comment by dharmit** [comment]
> I get that this feels somewhat nitpicky, but because of the sheer amount of files that are updated in charts PRs it makes it much easier for reviewers to give a review.

No, it doesn't sound nitpicky at all, TBH. Thanks a tonne for laying down the steps. I'll update the PR accordingly. 👍🏽

File: release.yaml
Comment (joshmeranda) [inline release.yaml]: We should avoid adding references to charts that weren't updated in this PR. Remove the `fleet-agent` and `rancher-vsphere-csi` fields please. Otherwise LGTM!

**Review by joshmeranda — APPROVED** [review]

**Review by joshmeranda — CHANGES_REQUESTED** [review]
see `release.yaml` comments above

File: release.yaml
Comment (dharmit) [reply release.yaml]: Well, those changes are based on `make check-release-yaml`.

**Review by joshmeranda — APPROVED** [review]
LGTM

**Comment by ericpromislow** [comment]
I don't think you need my review here -- reviews from Josh and Alexandre are sufficient.

**Comment by dharmit** [comment]
@joshmeranda @alexandreLamarre I have cherry-picked commits to be added on top of dev-v2.10 branch. Can you PTAL? I think the CI needs to be re-triggered to be run against 2.10 branch.

**Comment by mallardduck** [comment]
Seeing this upon install:

> W0813 13:11:43.161908      44 warnings.go:70] spec.template.spec.containers[2].ports[0]: duplicate port definition with spec.template.spec.containers[1].ports[0]

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml
Comment (mallardduck) [inline packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml]: Seems like this doesn't actually have a default value. Is there supposed to be some sort of helper or lookup with this as an override when set? Or is this required? If required, we should add it as an empty (null) value and put a note that it should be set or something.

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch
Comment (mallardduck) [inline packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch]: This is OK for now - because it seems like it should only need to be on or off. However, other `networkPolicy` related values in this file make it an object. So if we need more than one value for this I would conform to that example as well. I.e.

```yaml
networkPolicy:
  - enabled: false
  - namespace: null # user must set
  ```

**Review by mallardduck — CHANGES_REQUESTED** [review]
Not quite working out the box for me when I set to enabled. In the rancher installed apps section it thinks that it has network policies, but they 404 when I click their links.

**Comment by mallardduck** [comment]
> I don't think you need my review here -- reviews from Josh and Alexandre are sufficient.

Screenshot 2024-08-13 at 9 35 29 AM


@ericpromislow - I think that because you left a "Requested Changes" review you must provide an "Approved" review before this PR will be passable. Ignore mine since I added them today intentionally for changes I suggested. But notice your's and alex...I think big picture we need a new PR even after mine and yours are approved since Alex can review currently.

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch
Comment (dharmit) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch]: That was my initial thought when introducing this value. But decided against it as it's a single value.

**Comment by joshmeranda** [comment]
> Seeing this upon install:
> 
> > W0813 13:11:43.161908      44 warnings.go:70] spec.template.spec.containers[2].ports[0]: duplicate port definition with spec.template.spec.containers[1].ports[0]

@mallardduck This one is coming from upstream: https://github.com/grafana/helm-charts/blob/main/charts/grafana/templates/_pod.tpl#L1016-L1021

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch
Comment (joshmeranda) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/patch/values.yaml.patch]: Its a single value, but if we need to make it an object later, that could affect upgrade paths / is not backward compatible. Unless there is a specific reason this needs to be a single value, it would be better as an object.

Good catch @mallardduck

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml
Comment (dharmit) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml]: It's there in `values.yaml`.
https://github.com/rancher/charts/blob/4411cc0ccaca933bb9a495413dd887d0922e9033/charts/rancher-monitoring/104.1.0%2Bup57.0.3/values.yaml#L335-L347

File: release.yaml
Comment (dharmit) [inline release.yaml]: The changes in this file are a result of doing `make validate` and `make check-release-yaml`. I have made zero manual changes to this.

**Comment by dharmit** [comment]
> Not quite working out the box for me when I set to enabled. In the rancher installed apps section it thinks that it has network policies, but they 404 when I click their links.

@mallardduck IIUC (which is based on description in https://github.com/rancher/rancher/issues/45603), you have to open Prometheus Targets from the Monitoring view and check if `serviceMonitor/kube-system/rancher-monitoring-ingress-nginx` is up. I'm not sure where in the Installed Apps would you see a Network Policy because I see none on my RKE2 setup. Also, please set up an RKE2 with CIS enabled. Below are the configuration files I used:

```sh
$ cat /etc/rancher/rke2/config.yaml
cni: calico
write-kubeconfig-mode: "0644"
profile: cis
pod-security-admission-config-file: /etc/rancher/rke2/rke2-custom-pss.yaml

$ cat /etc/rancher/rke2/rke2-custom-pss.yaml
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: PodSecurity
  configuration:
    apiVersion: pod-security.admission.config.k8s.io/v1beta1
    kind: PodSecurityConfiguration
    defaults:
      enforce: "restricted"
      enforce-version: "latest"
      audit: "restricted"
      audit-version: "latest"
      warn: "restricted"
      warn-version: "latest"
    exemptions:
      usernames: []
      runtimeClasses: []
      namespaces: [kube-system, cis-operator-system, tigera-operator, cattle-system, cattle-fleet-system, cattle-fleet-local-system, cattle-monitoring-system]
```

File: release.yaml
Comment (joshmeranda) [reply release.yaml]: `charts-build-scripts` is really only checking that everything that is going to be released is included, it does not check that things are should not be released are not included. In that process it might add unexpected things to `release.yaml`

Comment (dharmit) [reply release.yaml]: IMO doing `make validate` is sorting stuff alphabetically. However, the YAML it's generating is broken. That is fixed by running `make check-release-yaml`. That's been my observation so far.

Comment (joshmeranda) [reply release.yaml]: Yeah that whole, `check-release-yaml` scripts stuff is annoying. If I remember correctly the yaml itself isn't "broken" just not in the desired format (whitespce vs no whitespace in array data).

> Well, those changes are based on make check-release-yaml.

`check-release.yaml` does not modify data, it basically just runs it through `yq` and checks if that changed release.yaml. Those changes are coming from `make validate` which does a whole lot more than just alphabetize `release.yaml` (although I can remember if thats done in `check-release-yaml` or `make validate`). `make validate` basically compares the local list of charts with those found in the last release, and reports on anything that should be in released and wasn't. If you wanna see the nitty gritty look here: https://github.com/rancher/charts-build-scripts/blob/master/pkg/validate/validate.go.

We should not be adding charts to `release.yaml` unless we made changes to that chart in the same PR. We've all been guilty of doing this before since they are easy to miss changes, so don't sweat it :)

Comment (dharmit) [reply release.yaml]: > check-release.yaml does not modify data, it basically just runs it through yq and checks if that changed release.yaml

The `scripts/check-release-yaml` has this `yq -i release.yaml` which modifies `release.yaml` in place. But you're right that it's coming from `make validate` in this case.

> make validate which does a whole lot more than just alphabetize release.yaml (although I can remember if thats done in check-release-yaml or make validate). make validate basically compares the local list of charts with those found in the last release, and reports on anything that should be in released and wasn't. 

Yep, makes sense. In case of this PR, it's alphabetizing as well as adding the rc chart to the `relesae.yaml`.

> We should not be adding charts to release.yaml unless we made changes to that chart in the same PR. We've all been guilty of doing this before since they are easy to miss changes, so don't sweat it :)

I'm lost. Should this PR have the changes to `release.yaml` or not? Anyway, let's discuss this over a call for quicker resolution.

Comment (joshmeranda) [reply release.yaml]: Realized this was my mistake getting confused / not looking carefully enough. Changes to / `release.yaml` were just reordering + adding monitoring charts versions.

File: packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml
Comment (mallardduck) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml]: So I think looking at other similar situations to this one, what we actually want to do here is:
1. define a new item in `_helpers.tpl` for this,
2. Call that helper here to reference that instead,
3. Allow the `values.yaml` to have the `namespaceOverride` value default be empty.

Comment (mallardduck) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml]: So in this file we would replicate:
https://github.com/dharmit/rancher-charts/blob/aa9a6b007735706c311588752309cda89139982f/charts/rancher-monitoring/104.1.1-rc1%2Bup57.0.3/charts/prometheus-node-exporter/templates/daemonset.yaml#L5C26-L5C60

But call it: `"rke2-ingress-nginx.namespace"`

Comment (mallardduck) [reply packages/rancher-monitoring/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/exporters/ingress-nginx/network-policy.yaml]: And in the helpers file, I think we would mimic something like:
https://github.com/dharmit/rancher-charts/blob/aa9a6b007735706c311588752309cda89139982f/charts/rancher-monitoring/104.1.1-rc1%2Bup57.0.3/templates/_helpers.tpl#L301-L310

Since we know the default value should generally be `kube-system`, similar to how the default there is `kube-state-metrics`.

**Comment by mallardduck** [comment]
Edit: Disregard original message for now - Prachi can dismiss the blocking review.
Will review the PR first thing tomorrow and review again.

**Comment by dharmit** [comment]
> Edit: Disregard original message for now - Prachi can dismiss the blocking review.


TBH, I have been backing up old branch and force pushing each time. So, in a sense, this PR keeps getting fresh each time. 😉 

> Will review the PR first thing tomorrow and review again.

Before you spend time reviewing, I want to get the thing about RC clarified. Last time I pinged you on Slack, we agreed upon following steps mentioned in https://github.com/rancher/charts/pull/4404#issuecomment-2344160459 for this PR. However, a recent Slack thread started by @joshmeranda has got me concluding that I don't need to do those and only have a PR with the actual changes. Did I get this right?

**Comment by mallardduck** [comment]
> I don't need to do those and only have a PR with the actual changes. Did I get this right?

Yes that's correct - with the clarification we got from @joshmeranda, there's no need for this to bump RC.1 to RC.2.

**Comment by dharmit** [comment]
Closing this PR as it looks like @jbiers got this merged via #4545. The process for opening PR kept changing, and I got embedded into a different team, which led to work going unmerged and unnoticed. Disappointed. 😞

**Pull request closed without merging**
